internals/plan/plan.go

```diff
@@ -924,16 +924,78 @@ func (p *Plan) Validate() error {
 // services to be properly started, in the order that they must be started.
 // An error is returned when a provided service name does not exist, or there
 // is an order cycle involving the provided service or its dependencies.
-func (p *Plan) StartOrder(names []string) ([]string, error) {
-	return order(p.Services, names, false)
+func (p *Plan) StartOrder(names []string) ([][]string, error) {
+	orderedNames, err := order(p.Services, names, false)
+	if err != nil {
+		return nil, err
+	}
+
+	return createLanes(orderedNames, p.Services)
 }
 
 // StopOrder returns the required services that must be stopped for the named
 // services to be properly stopped, in the order that they must be stopped.
 // An error is returned when a provided service name does not exist, or there
 // is an order cycle involving the provided service or its dependencies.
-func (p *Plan) StopOrder(names []string) ([]string, error) {
-	return order(p.Services, names, true)
+func (p *Plan) StopOrder(names []string) ([][]string, error) {
+	orderedNames, err := order(p.Services, names, true)
+	if err != nil {
+		return nil, err
+	}
+
+	return createLanes(orderedNames, p.Services)
+}
+
+func getOrCreateLane(currentLane int, service *Service, serviceLaneMapping map[string]int) int {
+	// if the service has been mapped to a lane
+	if lane, ok := serviceLaneMapping[service.Name]; ok {
+		return lane
+	}
+
+	// if any dependency has been mapped to a lane
+	for _, dependency := range service.Requires {
+		if lane, ok := serviceLaneMapping[dependency]; ok {
+			return lane
+		}
+	}
+
+	// neither the service itself nor any of its dependencies is mapped to an existing lane
+	return currentLane + 1
+}
+
+func mapServiceToLane(service *Service, lane int, serviceLaneMapping map[string]int) {
+	serviceLaneMapping[service.Name] = lane
+
+	// map the service's dependencies to the same lane
+	for _, dependency := range service.Requires {
+		serviceLaneMapping[dependency] = lane
+	}
+}
+
+func createLanes(names []string, services map[string]*Service) ([][]string, error) {
+	serviceLaneMapping := make(map[string]int)
+
+	// Map all services into lanes.
+	var lane = -1
+	for _, name := range names {
+		service, ok := services[name]
+		if !ok {
+			return nil, &FormatError{
+				Message: fmt.Sprintf("service %q does not exist", name),
+			}
+		}
+
+		lane = getOrCreateLane(lane, service, serviceLaneMapping)
+		mapServiceToLane(service, lane, serviceLaneMapping)
```

Comment:
It seems odd to me that we don't do this inside `getOrCreateLane`, because as it is, `getOrCreateLane` doesn't ever actually create the lane, right? Can't we get rid of `mapServiceToLane` and just have `getOrCreateLane` fall through to the code in `mapServiceToLane` in the bottom case when it needs to create it?

Reply:
You are right that `getOrCreateLane` doesn't really create the lane.

But if we move the logic of `mapServiceToLane` to the bottom case, there will be issues:

For example, if "svc1 depends on svc2, svc2 depends on svc3":

1. When handling svc1, it's not in the map, so both svc1 and svc2 are on the map.
2. When handling svc2, since it's already in the map, it will not reach the bottom case, hence svc3 will not be put in the map.
3. When handling svc3, since it's not in the map, a new lane will be created, while in fact it should be put in the same lane.

That's why I separated the logic of creating lanes and mapping service to lanes.

That said, We don't have to call `mapServiceToLane` in `createLanes`, but inside `getOrCreateLane`. Shall I change it?

Reply:
Yes, I think that (calling `mapServiceToLane` inside `getOrCreateLane`) would make more sense.

Reply:
I moved `mapServiceToLane` inside `getOrCreateLane`. I really don't like calling the function 3 times in 3 cases. I have two other versions:

The first version I created a bool like `laneAlreadyExist` and use conditions on it for the second/third case, but I find it slightly harder to read than simply return directly.

The second version I had a wrapper, something to this effect:

```go
func getOrCreateLane(currentLane int, service *Service, serviceLaneMapping map[string]int) int {
	// Check if the service has been mapped to a lane
	if lane, ok := getServiceLane(service, serviceLaneMapping); ok {
		return lane
	}

	// Check if any dependency has been mapped to a lane
	for _, dependency := range service.Requires {
		if lane, ok := getServiceLane(dependency, serviceLaneMapping); ok {
			return lane
		}
	}

	// Neither the service itself nor any of its dependencies is mapped to an existing lane
	lane := currentLane + 1
	mapServiceToLane(service, lane, serviceLaneMapping)
	return lane
}

func getServiceLane(name string, serviceLaneMapping map[string]int) (int, bool) {
	if lane, ok := serviceLaneMapping[name]; ok {
		mapServiceToLane(name, lane, serviceLaneMapping)
		return lane, true
	}
	return 0, false
}
```

But I think it's also harder to read.

After consideration, I kept the current version which is to call `mapServiceToLane` in each case.